src/test/integration/java/com/emt/service/MatchServiceIT.java

```diff
@@ -57,4 +64,74 @@ public void testEloRatingUpdateAfterMatch() {
     assertThat(updatedSecondPlayer.getEloRating()).isLessThan(initialRatingSecondPlayer);
   }
+
+  @Test
+  @Transactional
+  public void testCancelMatch_shouldRevertEloRatingsAndRemoveMatch() {
+    PlayerResponse winner =
+        playerService.createPlayer(CreatePlayerRequest.builder().nickname("Winner").build());
+    PlayerResponse loser =
+        playerService.createPlayer(CreatePlayerRequest.builder().nickname("Loser").build());
+
+    BigDecimal initialRatingWinner = winner.eloRating();
+
+    CreateMatchRequest matchRequest = new CreateMatchRequest(winner.playerId(), loser.playerId());
+    MatchResponse matchToCancel = matchService.createMatch(matchRequest);
+
+    matchService.cancelMatch(matchToCancel.matchId());
+
+    Player updatedWinner = playerService.getPlayerById(winner.playerId());
+
+    assertThat(updatedWinner.getEloRating().setScale(2, BigDecimal.ROUND_HALF_UP))
+        .isEqualTo(initialRatingWinner.setScale(2, BigDecimal.ROUND_HALF_UP));
+    assertThat(matchRepository.count()).isZero();
+  }
+
+  @Test
+  public void testCancelNonExistentMatch_shouldThrowMatchNotFoundException() {
+    Long nonExistentMatchId = 999L;
+    MatchNotFoundException exception =
+        assertThrows(
+            MatchNotFoundException.class, () -> matchService.cancelMatch(nonExistentMatchId));
+
+    assertThat(exception.getMessage()).contains("Match not found with id");
+  }
+
+  @Test
+  @Transactional
+  public void testCancelMatch_shouldRevertEloRatingsAndRemoveOneMatch() {
+    PlayerResponse[] players = new PlayerResponse[6];
+    for (int i = 0; i < players.length; i++) {
+      players[i] =
+          playerService.createPlayer(
+              CreatePlayerRequest.builder().nickname("Player" + (i + 1)).build());
+    }
+
+    BigDecimal[] initialRatings = new BigDecimal[players.length];
+    for (int i = 0; i < players.length; i++) {
+      initialRatings[i] = players[i].eloRating();
+    }
+
+    List<MatchResponse> createdMatches = new ArrayList<>();
+    for (int i = 0; i < players.length; i += 2) {
+      CreateMatchRequest matchRequest =
+          new CreateMatchRequest(players[i].playerId(), players[i + 1].playerId());
+      createdMatches.add(matchService.createMatch(matchRequest));
+    }
+
+    matchService.cancelMatch(createdMatches.get(1).matchId());
+
+    List<Match> subsequentMatches =
+        matchRepository.findMatchesByPlayersAfter(
+            createdMatches.get(1).createdAt(), players[2].playerId(), players[3].playerId());
+
+    for (Match match : subsequentMatches) {
+      if (match.getWinner().getPlayerId() == players[3].playerId()
+          || match.getLoser().getPlayerId() == players[3].playerId()) {
+        Player updatedPlayer = playerService.getPlayerById(players[3].playerId());
+        assertThat(updatedPlayer.getEloRating())
+            .isEqualTo(initialRatings[3].setScale(2, BigDecimal.ROUND_HALF_UP));
+      }
+    }
+  }
```

Comment:
Poor readability and test flow, the test lacks logical usefulness.
avoid using for-each loops and magic indexes

When implementin a new logic, you should try to cover as many different possible situations as you can imagine: which match in the sequence of all matches was canceled, whether there were any matches before/after the canceled one, and with which players (related or unrelated to the canceled match). If yes, verify that the ratings are reverted correctly for the entire chain of subsequent matches